plutus-core/executables/src/PlutusCore/Executable/AstIO.hs

```diff
@@ -0,0 +1,192 @@
+
+-- Read a binary-encoded file (eg, Flat-encoded PLC)
+getBinaryInput :: Input -> IO BSL.ByteString
+getBinaryInput StdInput         = BSL.getContents
+getBinaryInput (FileInput file) = BSL.readFile file
+
+loadPirASTfromFlat :: Flat a => AstNameType -> Input -> IO (PirProg a)
+loadPirASTfromFlat flatMode inp = do
+    input <- getBinaryInput inp
+    case flatMode of
+        Named -> handleResult $ unflat input
+        _     -> unsupportedNameTypeError flatMode
+    where
+      handleResult =
+          \case
+           Left e  -> error $ "Flat deserialisation failure: " ++ show e
+           Right r -> return r
+
+
+-- | Read and deserialise a Flat-encoded PIR/PLC AST
+loadPlcASTfromFlat :: Flat a => AstNameType -> Input -> IO (PlcProg a)
+loadPlcASTfromFlat flatMode inp = do
+    input <- getBinaryInput inp
+    case flatMode of
+        Named     -> handleResult $ unflat input
+        DeBruijn  ->
+            do
+              deserialised <- handleResult $ unflat input
+              let namedProgram = PLC.programMapNames fakeTyNameDeBruijn fakeNameDeBruijn deserialised
+              fromDeBruijn namedProgram
+        NamedDeBruijn ->
+            do
+              deserialised <- handleResult $ unflat input
+              fromDeBruijn deserialised
+    where
+      handleResult =
+          \case
+           Left e  -> error $ "Flat deserialisation failure: " ++ show e
+           Right r -> return r
+
+      fromDeBruijn :: PlcProgramNdB ann -> IO (PlcProg ann)
+      fromDeBruijn prog = do
+        case PLC.runQuote $
+             runExceptT @PLC.FreeVariableError $ traverseOf PLC.progTerm PLC.unDeBruijnTerm prog of
+          Left e  -> error $ show e
+          Right p -> return p
+
+-- | Read and deserialise a Flat-encoded UPLC AST
+loadUplcASTfromFlat :: Flat ann => AstNameType -> Input -> IO (UplcProg ann)
+loadUplcASTfromFlat flatMode inp = do
+    input <- getBinaryInput inp
+    case flatMode of
+        Named -> handleResult $ unflat input
+        DeBruijn -> do
+            deserialised <- handleResult $ unflat input
+            let namedProgram = UPLC.programMapNames fakeNameDeBruijn deserialised
+            -- ^ namedProgram has names that look like `FakeNamedDeBruijn "i" ix`, where
+            -- ix is the de Bruijn index.
+            fromDeBruijn namedProgram
+            -- ^ This converts the indices to Uniques so that we end up with proper names.
+        NamedDeBruijn -> do
+            deserialised <- handleResult $ unflat input
+            fromDeBruijn deserialised
+  where
+    fromDeBruijn :: UplcProgramNdB ann -> IO (UplcProg ann)
+    fromDeBruijn prog = do
+      case PLC.runQuote $
+           runExceptT @UPLC.FreeVariableError $ traverseOf UPLC.progTerm UPLC.unDeBruijnTerm prog of
+        Left e  -> error $ show e
+        Right p -> return p
+    handleResult =
+        \case
+            Left e  -> error $ "Flat deserialisation failure: " ++ show e
+            Right r -> return r
```

Comment:
`handleResult` is duplicated 2 times (3 in total)

Reply:
Thanks!  When I started to tidy that up I realised that there was a lot of monadic stuff in there that we didn't actually need.